AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/TgzPackageExtract.kt

```diff
@@ -0,0 +1,324 @@
+ * The file format of downloaded file is tgz, so I created this class to extract tgz file to provided directory.
+ *
+ * This file in kotlin is used for extracting tgz file.
+ * https://android.googlesource.com/platform/tools/tradefederation/+/master/src/com/android/tradefed/util/TarUtil.java
+ *
+ * Also added zip slip safety
+ * https://snyk.io/research/zip-slip-vulnerability
+ *
+ * Also extracted gzip files with file size limit 100MB and number of entries in the tar file
+ */
+
+class TgzPackageExtract(private val context: Context) {
+    private val GZIP_SIGNATURE = byteArrayOf(0x1f, 0x8b.toByte())
+    private var requiredMinSpace: Long = 0
+    private var availableSpace: Long = 0
+
+    val BUFFER = 512
+    val TOO_BIG_SIZE: Long = 0x6400000 // max size of unzipped data, 100MB
+    val TOO_MANY_FILES = 1024 // max number of files
+
+    var count = 0
+    var total: Long = 0
+    var data = ByteArray(BUFFER)
+
+    /**
+     * Determine whether a file is a gzip.
+     *
+     * @param file the file to check.
+     * @return whether the file is a gzip.
+     * @throws IOException if the file could not be read.
+     */
+    @Throws(IOException::class)
+    fun isGzip(file: File?): Boolean {
+        val signature = ByteArray(GZIP_SIGNATURE.size)
+        FileInputStream(file).use { stream ->
+            if (stream.read(signature) != signature.size) {
+                return false
+            }
+        }
+        return GZIP_SIGNATURE.contentEquals(signature)
+    }
+
+    /**
+     * Untar and ungzip a tar.gz file to a AnkiDroid/addons directory.
+     *
+     * @param tarballFile the .tgz file to extract
+     * @param addonsDir   the AnkiDroid addons directory
+     * @return the temp directory.
+     * @throws FileNotFoundException if .tgz file or ungzipped file i.e. .tar file not found
+     * @throws IOException
+     */
+    @Throws(Exception::class)
+    fun extractTarGzipToAddonFolder(tarballFile: File, addonsDir: File) {
+
+        require(isGzip(tarballFile)) { context.getString(R.string.not_valid_js_addon, tarballFile.absolutePath) }
+
+        if (!addonsDir.exists()) {
+            context.getString(R.string.could_not_create_dir, addonsDir.absolutePath)
+        }
+
+        // Make sure we have 2x the tar file size in free space (1x for tar file, 1x for unarchived tar file contents
+        requiredMinSpace = tarballFile.length() * 2
+        availableSpace = Utils.determineBytesAvailable(addonsDir.canonicalPath)
+        if (requiredMinSpace > availableSpace) {
+            Timber.e("Not enough space to extract file, need %d, available %d", requiredMinSpace, availableSpace)
+            throw IOException(context.getString(R.string.import_log_insufficient_space_error, Formatter.formatFileSize(context, requiredMinSpace), Formatter.formatFileSize(context, availableSpace)))
+        }
+
+        // If space available then unGZip it
+        val tarTempFile = unGzip(tarballFile, addonsDir)
+
```

Comment:
right here - this is where you should set deleteOnExit
This is a very very minor improvement in error handling, since you are already deleting it in a finally it should be good right? But, what if the system actually puts the app in the background while calculating size, or kills the app for memory pressure etc? Before you get to the finally? If you set deleteOnExit it will clean up even then!

```suggestion
        val tarTempFile = unGzip(tarballFile, addonsDir)
        tarTempFile.deleteOnExit()

```